Enable particle input from file

diehlpk opened the pull request:
Attempt to read particles from a CSV file.

Closes #9

streeve commented:
Updated here now that the crack branching problem is merged; also split into a new example to separate the specific file format from the source

diehlpk commented:
@streeve I tried to compile the new version of the example. 

1. I added the example to CMake to finally compile it 
2. I fixed some of the C++ compiler errors for reading the CSV file

However, I get some Kokkos errors while compiling.

```
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp: In function ‘void read_particles(std::__cxx11::string, ParticleType&)’:
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:56:14: error: there are no arguments to ‘slice_v’ that depend on a template parameter, so a declaration of ‘slice_v’ must be available [-fpermissive]
     auto v = slice_v();
              ^~~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:56:14: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:57:14: error: there are no arguments to ‘slice_f’ that depend on a template parameter, so a declaration of ‘slice_f’ must be available [-fpermissive]
     auto f = slice_f();
              ^~~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:58:17: error: there are no arguments to ‘slice_type’ that depend on a template parameter, so a declaration of ‘slice_type’ must be available [-fpermissive]
     auto type = slice_type();
                 ^~~~~~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:59:16: error: there are no arguments to ‘slice_rho’ that depend on a template parameter, so a declaration of ‘slice_rho’ must be available [-fpermissive]
     auto rho = slice_rho();
                ^~~~~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:60:14: error: there are no arguments to ‘slice_u’ that depend on a template parameter, so a declaration of ‘slice_u’ must be available [-fpermissive]
     auto u = slice_u();
              ^~~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:61:19: error: there are no arguments to ‘slice_nofail’ that depend on a template parameter, so a declaration of ‘slice_nofail’ must be available [-fpermissive]
     auto nofail = slice_nofail();
                   ^~~~~~~~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:66:41: error: expected primary-expression before ‘policy’
         Kokkos::RangePolicy<exec_space> policy( 0, x.size() ),
                                         ^~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp: In lambda function:
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:69:19: error: ‘pid’ was not declared in this scope
             pvol( pid ) = vol( i );
                   ^~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp: In function ‘int main(int, char**)’:
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:159:19: error: ‘class std::shared_ptr<CabanaPD::Particles<Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, CabanaPD::PMB> >’ has no member named ‘update_after_read’
         particles.update_after_read( exec_space(), halo_width, num_cell );
                   ^~~~~~~~~~~~~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp: In instantiation of ‘void read_particles(std::__cxx11::string, ParticleType&) [with ParticleType = std::shared_ptr<CabanaPD::Particles<Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, CabanaPD::PMB> >; std::__cxx11::string = std::__cxx11::basic_string<char>]’:
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:157:46:   required from here
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:49:61: error: no type named ‘memory_space’ in ‘class std::shared_ptr<CabanaPD::Particles<Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, CabanaPD::PMB> >’
     using memory_space = typename ParticleType::memory_space;
                                                             ^
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:54:25: error: ‘class std::shared_ptr<CabanaPD::Particles<Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, CabanaPD::PMB> >’ has no member named ‘slice_x’
     auto px = particles.slice_x();
               ~~~~~~~~~~^~~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:55:27: error: ‘class std::shared_ptr<CabanaPD::Particles<Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, CabanaPD::PMB> >’ has no member named ‘slice_vol’
     auto pvol = particles.slice_vol();
                 ~~~~~~~~~~^~~~~~~~~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:56:21: error: ‘slice_v’ was not declared in this scope
     auto v = slice_v();
              ~~~~~~~^~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:56:21: note: suggested alternative: ‘ssize_t’
     auto v = slice_v();
              ~~~~~~~^~
              ssize_t
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:57:21: error: ‘slice_f’ was not declared in this scope
     auto f = slice_f();
              ~~~~~~~^~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:57:21: note: suggested alternative: ‘ssize_t’
     auto f = slice_f();
              ~~~~~~~^~
              ssize_t
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:58:27: error: ‘slice_type’ was not declared in this scope
     auto type = slice_type();
                 ~~~~~~~~~~^~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:58:27: note: suggested alternative: ‘H5Rget_type’
     auto type = slice_type();
                 ~~~~~~~~~~^~
                 H5Rget_type
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:59:25: error: ‘slice_rho’ was not declared in this scope
     auto rho = slice_rho();
                ~~~~~~~~~^~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:60:21: error: ‘slice_u’ was not declared in this scope
     auto u = slice_u();
              ~~~~~~~^~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:60:21: note: suggested alternative: ‘ssize_t’
     auto u = slice_u();
              ~~~~~~~^~
              ssize_t
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:61:31: error: ‘slice_nofail’ was not declared in this scope
     auto nofail = slice_nofail();
                   ~~~~~~~~~~~~^~
/home/diehlpk/Compile/CabanaPD/examples/read_file.cpp:61:31: note: suggested alternative: ‘nofail’
     auto nofail = slice_nofail();
                   ~~~~~~~~~~~~^~
```

Could you please have a look?

streeve commented on examples/read_file.cpp:
```diff
@@ -0,0 +1,199 @@
+
+    auto px = particles.slice_x();
+    auto pvol = particles.slice_vol();
+    auto v = slice_v();
```
fixed

streeve commented:
@diehlpk I fixed the compilation locally - should be much more usable now

diehlpk commented:
@streeve 

I get now a new error which was not there before:

```
In file included from /home/diehlpk/Compile/CabanaPD/src/CabanaPD.hpp:16,
                 from /home/diehlpk/Compile/CabanaPD/examples/elastic_wave.cpp:19:
/home/diehlpk/Compile/CabanaPD/src/CabanaPD_Comm.hpp: In member function ‘void CabanaPD::HaloIds<LocalGridType, PositionSliceType>::build(const LocalGridType&)’:
/home/diehlpk/Compile/CabanaPD/src/CabanaPD_Comm.hpp:114:33: error: ‘getTopology’ is not a member of ‘Cajita’
         auto topology = Cajita::getTopology( local_grid );
                                 ^~~~~~~~~~~
/home/diehlpk/Compile/CabanaPD/src/CabanaPD_Comm.hpp:114:33: note: suggested alternative:
In file included from /home/diehlpk/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cabana-master-35llli4s5ak7e2at4sxta22swwr6ydzb/include/Cajita.hpp:34,
                 from /home/diehlpk/Compile/CabanaPD/src/CabanaPD_Boundary.hpp:18,
                 from /home/diehlpk/Compile/CabanaPD/src/CabanaPD.hpp:15,
                 from /home/diehlpk/Compile/CabanaPD/examples/elastic_wave.cpp:19:
/home/diehlpk/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cabana-master-35llli4s5ak7e2at4sxta22swwr6ydzb/include/Cajita_ParticleGridDistributor.hpp:57:1: note:   ‘Cajita::Impl::getTopology’
 getTopology( const LocalGridType& local_grid )
 ^~~~~~~~~~~
In file included from /home/diehlpk/Compile/CabanaPD/src/CabanaPD.hpp:16,
                 from /home/diehlpk/Compile/CabanaPD/examples/elastic_wave.cpp:19:
/home/diehlpk/Compile/CabanaPD/src/CabanaPD_Comm.hpp:115:40: error: ‘getUniqueTopology’ is not a member of ‘Cabana’
         auto unique_topology = Cabana::getUniqueTopology(
                                        ^~~~~~~~~~~~~~~~~
/home/diehlpk/Compile/CabanaPD/src/CabanaPD_Comm.hpp:115:40: note: suggested alternative:
In file included from /home/diehlpk/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cabana-master-35llli4s5ak7e2at4sxta22swwr6ydzb/include/Cabana_Distributor.hpp:20,
                 from /home/diehlpk/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cabana-master-35llli4s5ak7e2at4sxta22swwr6ydzb/include/Cabana_Core.hpp:40,
                 from /home/diehlpk/Compile/CabanaPD/src/CabanaPD_Boundary.hpp:17,
                 from /home/diehlpk/Compile/CabanaPD/src/CabanaPD.hpp:15,
                 from /home/diehlpk/Compile/CabanaPD/examples/elastic_wave.cpp:19:
/home/diehlpk/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cabana-master-35llli4s5ak7e2at4sxta22swwr6ydzb/include/Cabana_CommunicationPlan.hpp:360:25 note:   ‘Cabana::Impl::getUniqueTopology’
 inline std::vector<int> getUniqueTopology( MPI_Comm comm,
                         ^~~~~~~~~~~~~~~~~
In file included from /home/diehlpk/Compile/CabanaPD/src/CabanaPD.hpp:16,
                 from /home/diehlpk/Compile/CabanaPD/examples/elastic_wave.cpp:19:
/home/diehlpk/Compile/CabanaPD/src/CabanaPD_Comm.hpp: In constructor ‘CabanaPD::Comm<ParticleType, CabanaPD::PMB>::Comm(ParticleType&, int)’:
/home/diehlpk/Compile/CabanaPD/src/CabanaPD_Comm.hpp:207:33: error: ‘getTopology’ is not a member of ‘Cajita’
         auto topology = Cajita::getTopology( *local_grid );
                                 ^~~~~~~~~~~
/home/diehlpk/Compile/CabanaPD/src/CabanaPD_Comm.hpp:207:33: note: suggested alternative:
In file included from /home/diehlpk/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cabana-master-35llli4s5ak7e2at4sxta22swwr6ydzb/include/Cajita.hpp:34,
                 from /home/diehlpk/Compile/CabanaPD/src/CabanaPD_Boundary.hpp:18,
                 from /home/diehlpk/Compile/CabanaPD/src/CabanaPD.hpp:15,
                 from /home/diehlpk/Compile/CabanaPD/examples/elastic_wave.cpp:19:
/home/diehlpk/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-8.5.0/cabana-master-35llli4s5ak7e2at4sxta22swwr6ydzb/include/Cajita_ParticleGridDistributor.hpp:57:1: note:   ‘Cajita::Impl::getTopology’
 getTopology( const LocalGridType& local_grid )
 ^~~~~~~~~~~
make[2]: *** [examples/CMakeFiles/ElasticWave.dir/build.make:76: examples/CMakeFiles/ElasticWave.dir/elastic_wave.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:175: examples/CMakeFiles/ElasticWave.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

streeve commented:
Sorry, I forgot to say you'll need to update to the latest Cabana master (readme has been updated with the minimum commit necessary)

diehlpk commented:
@streeve The code compiles now, but we get a segmentation fault after reading the CSV file.

I nailed it down between  https://github.com/ORNL/CabanaPD/blob/8fb8f429d9d07c8ea2154c733ad641b98d2f6fe5/examples/read_file.cpp#L43 a and https://github.com/ORNL/CabanaPD/blob/8fb8f429d9d07c8ea2154c733ad641b98d2f6fe5/examples/read_file.cpp#L84.

streeve commented:
If you attach an example file I can test as well

diehlpk commented:
This is the CSV file and [easy.csv](https://github.com/ORNL/CabanaPD/files/12662800/easy.csv)n and the image the corresponding mesh

![mesh](https://github.com/ORNL/CabanaPD/assets/3922435/c0503790-f9b6-4c4a-a195-d11ad97b4494)

streeve commented:
@diehlpk as you can see from the last commit, there were a lot of necessary changes. Runs for me now locally

diehlpk commented:
@streeve thanks. I will try to run a larger geometry and apply the boundary conditions.

diehlpk commented:
@streeve How do I pass command line arguments?

I assume I pass` inputs.input_file` using the command line? 

For now, I hard-coded the filename for testing.

I added some plane to apply the boundary conditions for a refined mesh


```
   CabanaPD::RegionBoundary plane1( 0 , 1, 9, 10,
                                         -1, 1 );
 
          std::vector<CabanaPD::RegionBoundary> planes = { plane1};
          auto bc =
             createBoundaryCondition( CabanaPD::ForceCrackBranchBCTag{},
                                      exec_space{}, *particles, planes, 1.0 );
```

after running the simulation, the force is applied to the plane1 but not propagated to the other particles. Please take a look at the attached image.

![forces](https://github.com/ORNL/CabanaPD/assets/3922435/f2877b2f-4761-477f-b22b-7e6f5a53e211)

I attached the new CSV file [here](https://github.com/ORNL/CabanaPD/files/12811192/easy.csv).

Once, I know how to pass the file name, I can remove the hard-coded file name and commit my changes.

streeve commented:
> How do I pass command line arguments?
> 
> I assume I pass` inputs.input_file` using the command line?

Sorry, forgot to mention that it currently requires a flag `./exe -i input.file`

By the way, in #44 we're looking at including an input file name as an optional json input


I can take a look at the BC if you say a bit more on what the intended conditions are

diehlpk commented:
@streeve I committed my changes and I added a new [plane](https://github.com/diehlpk/CabanaPD/blob/f7f2ce18324a4f5fb0c0e7e4e84039b97eaa7bc4/examples/read_file.cpp#L147) and want to pull this layer with a force of 1 in y-direction. This is not the final boundary condition for the experiment. There, we would need to clamp the bottom part too.  

However, we first would need to allow for variable volumes and an arbitrary geometry.

This is more a sanity check.

diehlpk commented:
@streeve I fixed the issue with the force propagation, and there is now strain energy.

![good_strain](https://github.com/ORNL/CabanaPD/assets/3922435/6b789533-49b3-49fa-a1c7-11d4f8f6189b)

So I think reading the CSV file and applying the boundary conditions works now.

diehlpk commented:
@streeve I was running the new code version, and now all values for the force or displacement are `nan`. I told Paraview to plot all nan numbers in pink. 

![all-nan](https://github.com/ORNL/CabanaPD/assets/3922435/a3de7ec6-810b-4f4b-88a4-0eee57d67453)

diehlpk commented:
I tried to center the geometry at (0,0) and still get the same result. 

I pushed the new code and here is the new [csv](https://github.com/ORNL/CabanaPD/files/12835100/center.csv) file.

streeve commented:
@diehlpk I believe I fixed this by reverting to a 3d (pseudo-2d) system. Neighbor lists still need to be updated for true 2d support

diehlpk commented:
@streeve Thanks for the fix. It looks a little bit better, but the values in the regions with the boundary conditions are still not a number values, see the attached image.

![nan-top-bottom](https://github.com/ORNL/CabanaPD/assets/3922435/1c0a7a2f-72ca-4590-a424-fed6d591b5f7)

streeve commented:
To do:
- Generalize boundary for fields other than forces
- Remove dx (build a grid with the cutoff distance and fix halo=1)

streeve commented:
@diehlpk made the changes we discussed - let me know if you have any issues or if you have larger files for me to test with

diehlpk commented:
@streeve thanks, I will upload the coarsest grid, we used in the previous simulations this evening.

diehlpk commented:
@streeve Please find attached a mesh more suitable for the comparison with the experiment.

![new-mesh](https://github.com/ORNL/CabanaPD/assets/3922435/cb9efc94-5054-4083-aa5a-839370715f34)

Here is the file [astm-d638.csv](https://github.com/ORNL/CabanaPD/files/13233408/astm-d638.csv)

If that mesh works, we would be good to setup the experimet.

diehlpk commented:
@streeve mayve we should add a very small CSV file to the repo to execute the example. What do you think?

streeve commented:
> mayve we should add a very small CSV file to the repo to execute the example. What do you think?

I'm okay with that as long as it's very small (a dog-bone with 100s of particles would be nice)

streeve commented:
I ran with the astm file and can correctly create the BC we want, but I'm not sure if the issues I see locally are because of input values or a bug

streeve commented:
@diehlpk should be ready to test. The last commit adds the linear ramp for BC (as well as the option to turn off the BC at some point in the simulation). I also had to remove some duplicate positions with the test mesh I was using

streeve commented:
@diehlpk let me know if you notice any issues - I can try to fix before we meet next

diehlpk commented:
@streeve I was attending SC. I will testbit before the meeting back home.

diehlpk commented:
@streeve I prepared a 3D mesh and uploaded the mesh to the shared folder.

I will try this week to setup the 2d case with the experimental data.

Could you please try the 3d mesh?
